lib/src/modules/download_file.cc

```diff
@@ -113,31 +119,71 @@ namespace Modules {
+    return fail_response;
+  }
+
   // DownloadFile overrides callAction from the base BoltModule class since there's no need to run
   // any commands with DownloadFile. CallAction will simply download the file and return a result
   // based on if the download succeeded or failed.
   ActionResponse DownloadFile::callAction(const ActionRequest& request)
   {
     auto file_params = request.params();
     auto files = file_params.get<std::vector<lth_jc::JsonContainer>>("files");
-    auto destination = fs::path(files[0].get<std::string>("destination"));
     const fs::path& results_dir = request.resultsDir();
 
     ActionResponse response { ModuleType::Internal, request };
     lth_jc::JsonContainer result;
-    try {
-      Util::downloadFileFromMaster(master_uris_,
-                    file_download_connect_timeout_,
-                    file_download_timeout_,
-                    client_,
-                    module_cache_dir_->createCacheDir(files[0].get<std::string>("sha256")),
-                    destination,
-                    files[0]);
-      response.output = write_download_results(results_dir, EXIT_SUCCESS, "downloaded", "");
-    } catch (Module::ProcessingError& e) {
-      LOG_ERROR("Failed to download {1}; {2}", destination, e.what());
-      response.output = write_download_results(results_dir, EXIT_FAILURE, "", lth_loc::format("Failed to download {1}; {2}", destination, e.what()));
+    for (auto this_file : files) {
+      auto destination = fs::path(this_file.get<std::string>("destination"));
+      auto file_type = this_file.get<std::string>("file_type");
+      if (file_type == "file") {
+        try {
+          if (!fs::exists(destination.parent_path())) {
+            Util::createDir(destination.parent_path());
+          }
+          Util::downloadFileFromMaster(master_uris_,
+                        file_download_connect_timeout_,
+                        file_download_timeout_,
+                        client_,
+                        module_cache_dir_->createCacheDir(this_file.get<std::string>("sha256")),
+                        destination,
+                        this_file);
+        } catch (Module::ProcessingError& e) {
+          return failure_response(request, results_dir, lth_loc::format("Failed to download {1}; {2}", destination, e.what()));
+        }
+      } else if (file_type == "directory"){
+        if (fs::exists(destination)) {
+          if (!fs::is_directory(destination)) {
+            return failure_response(request, results_dir, lth_loc::format("Destination {1} already exists and is not a directory!", destination));
+          }
+        } else {
+          try {
+            Util::createDir(destination);
+          } catch (fs::filesystem_error& e) {
+            return failure_response(request, results_dir, lth_loc::format("Failed to create directory {1}; {2}", destination, e.what()));
+          }
+        }
+      } else if (file_type == "symlink") {
```

Comment:
Have we confirmed how Bolt and Puppet behave for symlinks (on POSIX and Windows) so we can be consistent?

Reply:
There's an acceptance test that uses puppet to check that puppet is reporting the symlink as:

`ensure => link`

So I _think_ they behave the same.